### PR TITLE
Setup project with standalone benchmarks

### PR DESCRIPTION
**Pull request by dlam**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        // in the individual module build.gradle files
+    }
+}
+
+allprojects {
+    repositories {
+        google()
+        jcenter()
+        maven {
+            url '/usr/local/google/home/dustinlam/google/androidx-master-dev/out/host/gradle/frameworks/support/build/support_repo'
```
Comment (ChrisCraik) [inline build.gradle]: use a local path, e.g. "./android-performance/temporary_repo"

File: benchmark/src/androidTest/java/com/example/benchmark/AutoBoxingBenchmark.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import androidx.benchmark.BenchmarkRule
+import androidx.test.filters.LargeTest
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+@LargeTest
+@RunWith(JUnit4::class)
+class AutoBoxingBenchmark {
```
Comment (ChrisCraik) [inline benchmark/src/androidTest/java/com/example/benchmark/AutoBoxingBenchmark.kt]: I did some experimenting with this one locally - not really happy with it yet. Let's remove this file for now.

If I messed around with the structure, it seemed I could get the VM to reuse boxed integers and make performance very similar (or worse, have one run pollute the next), e.g.:

    @Test
    fun integerArtCacheAlloc() {
        benchmarkRule.measure {
            (0..99).map { Integer(it) }
        }
    }

    @Test
    fun integerAlloc() {
        benchmarkRule.measure {
            (1000..1099).map { Integer(it) }
        }
    }

File: benchmark/src/androidTest/java/com/example/benchmark/BitmapBenchmark.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    @get:Rule
+    val benchmarkRule = BenchmarkRule()
+
+    private val context: Context = InstrumentationRegistry.getTargetContext()
+
+    @Test
+    fun bitmapGetPixelBenchmark() {
+        val inputStream = context.assets.open(JETPACK)
+        val bitmap = BitmapFactory.decodeStream(inputStream)
```
Comment (ChrisCraik) [inline benchmark/src/androidTest/java/com/example/benchmark/BitmapBenchmark.kt]: extract bitmap out of each test

File: benchmark/src/androidTest/java/com/example/benchmark/BitmapBenchmark.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        const val JETPACK = "images/jetpack.png"
+    }
+
+    @get:Rule
+    val benchmarkRule = BenchmarkRule()
+
+    private val context: Context = InstrumentationRegistry.getTargetContext()
+
+    @Test
+    fun bitmapGetPixelBenchmark() {
```
Comment (ChrisCraik) [inline benchmark/src/androidTest/java/com/example/benchmark/BitmapBenchmark.kt]: Avoid the name 'benchmark' in the test methods - it's redundant with the name in the class.

(For a long time, pre-JUnit4, we wrote every test starting with the name "test". dark times)

File: benchmark/src/androidTest/java/com/example/benchmark/JniBenchmark.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+@LargeTest
+@RunWith(JUnit4::class)
+class JniBenchmark {
+
+    @get:Rule
+    val benchmarkRule = BenchmarkRule()
+
+    @Test
+    fun nativeBenchmark() {
```
Comment (ChrisCraik) [inline benchmark/src/androidTest/java/com/example/benchmark/JniBenchmark.kt]: remove name 'benchmark' from these test methods

File: .idea/codeStyles/Project.xml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<component name="ProjectCodeStyleConfiguration">
```
Comment (ChrisCraik) [inline .idea/codeStyles/Project.xml]: can you copy the .gitignore content from android-architecture-components? that will remove everything in .idea

note also that you'll need to copy the ones in the project subdirectories

File: benchmark/.gitignore [hunk 7]
```diff
@@ -0,0 +1 @@
+/build
```
Comment (ChrisCraik) [inline benchmark/.gitignore]: copy from android-architecture-components, there should be more

File: benchmark/build.gradle [hunk 8]
```diff
@@ -0,0 +1,33 @@
+apply plugin: 'com.android.library'
```
Comment (ChrisCraik) [inline benchmark/build.gradle]: move the benchmark project into a top level directory called "BenchmarkSample"

That'll give us room to do more, separate studio projects.

File: benchmark/.gitignore (hunk 7 above)
Comment (dlam) [reply benchmark/.gitignore]: :+1:

File: benchmark/src/androidTest/java/com/example/benchmark/AutoBoxingBenchmark.kt (hunk 2 above)
Comment (dlam) [reply benchmark/src/androidTest/java/com/example/benchmark/AutoBoxingBenchmark.kt]: Isn't the point to write code in a way that doesn't take advantage of this and show that it doesn't though?

Comment (dlam) [reply benchmark/src/androidTest/java/com/example/benchmark/AutoBoxingBenchmark.kt]: What if I added this test case in here and named it something like `integerJvmReuseAlloc`?

File: build.gradle (hunk 1 above)
Comment (dlam) [reply build.gradle]: :+1:

File: .idea/codeStyles/Project.xml (hunk 6 above)
Comment (dlam) [reply .idea/codeStyles/Project.xml]: :+1:

File: benchmark/src/androidTest/java/com/example/benchmark/BitmapBenchmark.kt (hunk 4 above)
Comment (dlam) [reply benchmark/src/androidTest/java/com/example/benchmark/BitmapBenchmark.kt]: :+1:

File: benchmark/src/androidTest/java/com/example/benchmark/JniBenchmark.kt (hunk 5 above)
Comment (dlam) [reply benchmark/src/androidTest/java/com/example/benchmark/JniBenchmark.kt]: :+1:

File: benchmark/src/androidTest/java/com/example/benchmark/BitmapBenchmark.kt (hunk 3 above)
Comment (dlam) [reply benchmark/src/androidTest/java/com/example/benchmark/BitmapBenchmark.kt]: :+1:

File: .gitignore [hunk 9]
```diff
@@ -1,14 +1,38 @@
-*.iml
+/unbundled-build
```
Comment (ChrisCraik) [inline .gitignore]: this looks a lot more complex than the arch ones:

https://github.com/googlesamples/android-architecture-components/blob/master/.gitignore
https://github.com/googlesamples/android-architecture-components/blob/master/BasicSample/.gitignore

Comment (dlam) [reply .gitignore]: Ah.. I took it from androidx. I guess we can just ignore the whole .idea folder in the samples

**Pull request merged**